File: largest_odd_num_string.py
# 1903. Largest Odd Number in String

"""
given a string 'num' representing a large integer. return the largest-valued
odd integer as a string is a non-empty substring of 'num', or an empty string
"" if no odd integer exists. a substring is a contiguous sequence of
characters within a string
"""


class Solution(object):
    def largestOddNumber(self, num):
        """
        :type num: str
        :rtype: str
        """
        for i in range(len(num) - 1, -1, -1):
            if num[i] in {"1", "3", "5", "7", "9"}:
                return num[: i + 1]
        return ""


if __name__ == "__main__":
    obj = Solution()
    print(obj.largestOddNumber("52"))  # expect: 5
    print(obj.largestOddNumber("4206"))  # expect: ""
    print(obj.largestOddNumber("35427"))  # expect: 35427

File: largest_odd_num_string.c
// 1903. Largest Odd Number in String
#include "leetcode.h"

/*
 * given a string 'num' representing a large integer. return the largest-valued
 * odd integer as a string is a non-empty substring of 'num', or an empty string
 * "" if no odd integer exists. a substring is a contiguous sequence of
 * characters within a string
 */

char *largestOddNumber(char *num) {
  int n = strlen(num);
  for (int i = n - 1; i >= 0; i--)
    if ((num[i] - '0') % 2 == 1) {
      num[i + 1] = '\0';
      return num;
    }
  return "";
}

int main() {
  char *n1 = "52", *n2 = "4206", *n3 = "35427";
  printf("%s\n", largestOddNumber(n1)); // expect: 5
  printf("%s\n", largestOddNumber(n2)); // expect: NULL
  printf("%s\n", largestOddNumber(n3)); // expect: 35427
}
